### ISSUE TITLE
Doc request: CA scale-down and taint "ToBeDeletedByClusterAutoscaler"

### ISSUE THREAD

**Issue by thockin** [kubernetes/autoscaler#5201]
**Which component are you using?**:

Cluster Autoscaler

**Describe the solution you'd like.**:

A clear doc.

I am trying to comprehend the meaning of the ToBeDeletedByClusterAutoscaler taint with regards to services with load-balancers. 
 I know of https://github.com/kubernetes/kubernetes/pull/105946 which made this taint influence the configuration of which nodes are chosen to serve as dataplane for a given service.  We're looking at making this more elegant (https://github.com/kubernetes/enhancements/pull/3460) and I realize I can't find any authoritative doc on this taint.

Specifically:

1) Is this taint only applied to already-drained nodes?  Or will it be applied to nodes which are still serving pods?

2) Is this taint always set along with `unschedulable`, or can a node still be schedulable AND taineted?

3) Is this a reversible decision, or is it safe to assume that once this taint is applied, that node *is* going away?

4) Is there a clear lifecycle diagram or state machine for which actions CA takes at which points, and what assertions can be made?

What I am looking for is something like:

t0: Node N is running with K pods
t1: CA chooses N to scale down
t2: CA sets node N as `unschedulable`
t3: CA drains pods from node N
t4: Drain is done
t5: CA sets node N taint `ToBeDeletedByClusterAutoscaler`
t6: CA deletes node N in k8s
t7: CA deletes the VM corresponding to node N

What we are trying to do with LBs is twofold:

1) Avoid re-programming the sets of nodes on an LB when we don't need to
2) Introduce an intermediate LB state between "in set" (which retains current connections and allows new ones) and "out of set" (which breaks existing connections and disallows new ones).  This new state would retain current connections and disallow new ones.

To achieve this, we want to leave nodes in the LB set as long as possible and rely on LB->node healthchecks.  If this taint is set too eagerly, that could devolve into the the current behavior where nodes get no drain time.

Happy to discuss more if there are questions, but a good drawing would go a long way. :)

**Comment by thockin** [comment]
cc @alexanderConstantinescu FYI

**Comment by thockin** [comment]
Based on https://github.com/kubernetes/autoscaler/blob/8dcd3b3eb2d1d28b7ba05a3af48a171cb40547b9/cluster-autoscaler/core/scaledown/actuation/actuator.go#L92-L94 I suspect this taint is set "eagerly" (before draining) which says to me that we should NOT respect it as a trigger to remove a node from the LB set.

That said, I can't find where the k8s node is deleted (??), and I am afraid that what will happen is that the CA will try to delete the VM, but that will fail because the VM is still in the LB.  

Which does CA delete first - the underlying VM or the k8s Node?

If it deletes the VM first, we need a clear signal for everyone to drop references to the VM.  We have a label for LBs: `node.kubernetes.io/exclude-from-external-load-balancers: "true"`.  The taint is more general than that, but if you set the taint to early you hurt our LB draining. effort

**Comment by alexanderConstantinescu** [comment]
I +1 the idea @thockin had above, i.e: the CA can continue using `ToBeDeletedByClusterAutoscaler` and setting it in t3. If we however agree that the CA can set `node.kubernetes.io/exclude-from-external-load-balancers: "true"` at t5.5, then we'd simplify the service controller's logic and allow connection draining.

What concerns me with changing this sequence of events, is: the importance of possible transient failure modes in the CA / service controller. Here we are discussing one in the CA (i.e: the problem of deleting an instance which might still be referenced by LBs). On the KEP, I've mentioned another: https://github.com/kubernetes/enhancements/pull/3460/files#diff-8b43694810d37cff93222d7ae4286d4aed520f445f96ae13afba684554a98ce4R213-R235  

Transient failures should be non-events in Kube, but people seem to have tried to avoid them in the past and possibly for good reasons?

**Comment by thockin** [comment]
> If we however agree that the CA can set node.kubernetes.io/exclude-from-external-load-balancers: "true" at t5.5, then we'd simplify the service controller's logic and allow connection draining.

My concern is that this annotation is VERY SPECIFIC and if there should ever be something else that needs to wake up at that point in the sequence, it would not make a lot of sense to look for this signal.

**Comment by thockin** [comment]
@mwielgus This is the one I mentioned at kubecon

**Comment by thockin** [comment]
Revisiting this - it's blocking progress on the goal.

I propose that we add a new taint or annotation (whichever is preferred by CA?), something like "NodeDeletionImminent" which we set between t5 and t6 above.

To make this safe we need to confirm that CA's `DeleteNodeFromCloudProvider()` will be retried "soon" if the underlying VM can't be deleted yet.  I could not convince myself of that by code inspection, and would appreciate help.

@x13n

**Comment by x13n** [comment]
Another taint/annotation is doable, one my worry would be extra QPS to kube-apiserver in large clusters. Nowadays with priority&fairness probably less of a concern, but something to keep in mind.

However, I don't think CA can guarantee VM termination. If we first say NodeDeletionImminent via annotation and then get quota/permission/other - not necessarily retryable - issue causing cloud provider delete operation to fail (or CA crashes before it even makes the attempt), node may get stuck with it unless CA has a way of recovering from this.

Regarding retries: if a VM fails to be deleted, CA will quite likely consider it for deletion again in the following loop iterations though, as the node is empty (after drain) and tainted (so it stays empty): a perfect scale down candidate.

**Comment by thockin** [comment]
Good point - I doubt the QPS from this matters much - the total number of nodes and number of all-node watchers is pretty low.

So adding a new "node deletion is imminent" annotation at t5.5 seems like not a lot of work technically but ...

The deletion-failure is a real issue that will need to be addressed.  If LB controller defers its work until "node deletion is imminent", then I expect that `DeleteNodeFromCloudProvider()` will get errors.  @x13n explained that CA doesn't exactly "retry the deletion" but repeats the whole CA loop and (usually, hopefully) picks to delete the same node again, but maybe not!

I think that once "node deletion is imminent", we should not reverse that decision.  We would want to actually retry the delete and, if CA should crash-restart, recognize the state and follow-through with it.

This is the problem now - that's not such an obvious bit of work (it might be as easy as looking for a new annotation each loop, but that's not obvious) and it's not clear who would take it on.  @x13n says he can guide but doesn't have cycles himself to implement it in the very near term.  @alexanderConstantinescu is this something you'd be open to following up on?  Sometimes you go where the bugs lead you :)

I also felt like there's a deeper question I don't understand, which maybe @x13n can help me learn:  We already have a signal that "deletion is imminent and irreversible" - `deletionTimestamp`.  Would it be a viable design to have CA install a finalizer on the Node resource and then delete it?  I know the deletion is done by node-controller today, but it feels kind of dirty to add another CA-specific decoration that literally means the same thing.  Is there a full node-lifecycle documented somwhere?

Either way, the LB-set optimizations are stuck until we do something.  If we treat the existing `ToBeDeletedByClusterAutoscaler` taint as a signal to remove a node from LB sets, then CA just doesn't work with LB node draining.  If we don't treat `ToBeDeletedByClusterAutoscaler` as a signal to remove a node from LB sets, then LB controller will never do that, and VM deletion will perma-fail.

I'm going to re-title this issue.

**Comment by alexanderConstantinescu** [comment]
I am bit concerned that the node deletion process, today, is implicitly correctly ensured by the fact that the `ToBeDeletedByClusterAutoscaler ` taint is placed at t3 - with a time consuming operation succeeding it (the drain). Our discussion here only _assumes_ that the cloud provider will return an error if the VM is still referenced by some other construct the moment it's deleted. We don't really know, do we? By placing the signal to the service controller at t5.5, I get the feeling we increase the risk of having the load balancer reference un-reaped - because the VM got deleted before the service controller even got a chance to try. I am pretty sure the service controller, at the very least, performs one GET call for the VM when trying to remove the LB references to it. 

As to make this entire process explicit, we would need a signal in the other direction, i.e: the service controller which indicates that it has removed all references to the VM, allowing the CA to delete the VM. And that feels like a lot more work. A finalizer might be a step in the right direction, but it doesn't ensure that all controllers act in any particular coordinated order. If both controllers place a finalizer on the Node object, then I guess the CA can inspect the service controller's finalizer before deciding on if it should delete the VM. It would trigger the deletion of the Node in Kube, wait for the service controller to react and delete the finalizer once all references are removed, then delete the VM, and only then delete its own finalizer, feels strange....and given that the CA doesn't even delete the Node in Kube currently, it doesn't even feel straightforward.    

>  @x13n explained that CA doesn't exactly "retry the deletion" but repeats the whole CA loop and (usually, hopefully) picks to delete the same node again, but maybe not!

> This is the problem now - that's not such an obvious bit of work (it might be as easy as looking for a new annotation each loop, but that's not obvious

It seems (logically, at least) straightforward to add an annotation to the node object, such as: `ca-deletion-timestamp: ${SOME_TIMESTAMP}` which is added on the node object at 5.5 when the CA needs to indicate "node deletion is imminent". The CA can then sorts all nodes to be deleted in each sync, as to ensure a deterministic deletion order at all times - even between multiple nodes, crashes / restarts and sync loops?

> @alexanderConstantinescu is this something you'd be open to following up on? Sometimes you go where the bugs lead you :)

Sure, this year I am getting some dedicated time from my employer for upstream work, so I can probably assist on this topic. 

In any case: I would like to start working on some E2E tests for all of these LB optimizations we have in mind, so that we have some tangible indicators of progress. I.e: an E2E tests which fails before we start doing any work, and gets fixed once we have all pieces in place....maybe it's a bit too optimistic to have something as deterministic as an E2E, for something like this? 

So maybe the sequence of events should be:

1. Document the problem and intended fix on the KEP
2. Agree on the KEP 
3. (in parallel) Write some E2E tests which highlight the connectivity problems mentioned on the KEP
4. Implement changes
5. Validate that tests pass

**Comment by x13n** [comment]
If we use an annotation, CA will have to remember which annotated nodes it already deleted and avoid retrying VM deletion after a successful one. This state can be kept by CA in memory, but then it will be lost if CA crashes. Alternatively, CA could reflect successful VM deletion by modifying Node object again (e.g. dropping the annotation or adding another one), in which case it is semantically equivalent to having a finalizer, so just having a finalizer seems reasonable to me.

Alternatively, instead of removing the finalizer, CA could try to delete the entire Node object itself, but then we'd need to make sure node controller isn't going to be unhappy about this.

> Is there a full node-lifecycle documented somwhere?

No idea, I'd be interested to see such doc if it exists.

**Comment by MaciekPytel** [comment]
> Good point - I doubt the QPS from this matters much - the total number of nodes and number of all-node watchers is pretty low.

Mutating QPS to api-server (mostly needed for tainting) is the bottleneck for empty node deletion throughput in CA in the environments I'm familiar with. Of course this may be just because in default configuration we set the client side QPS limit too low. 

>  but repeats the whole CA loop and (usually, hopefully) picks to delete the same node again, but maybe not!

I think it's much worse than that. In default configuration CA will completely block scale-down for [3 minutes](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/main.go#L104) after any failure. 
Even if that particular feature is disabled, I think CA is much more likely to pick a different node than to retry the same one. The reason is that we only delete node that has been a scale-down candidate continously for 10 minutes (in default configuration) and at least in legacy scale-down ("legacy" scale-down is default so far, the new scale-down is WIP) we [reset the timer on any scale-down attempt](https://github.com/kubernetes/autoscaler/blob/1e1e88d6e15fc27b67b631a8c7d3a53c4c99f3c9/cluster-autoscaler/core/scaledown/legacy/legacy.go#L311). 
This means that CA will not retry the same node for 10 minutes (and it will [remove the taint](https://github.com/kubernetes/autoscaler/blob/1e1e88d6e15fc27b67b631a8c7d3a53c4c99f3c9/cluster-autoscaler/core/scaledown/actuation/actuator.go#L426) on failure). 

To sum up: I don't think CA today is anywhere near being able to handle failures to remove nodes gracefully and my guess is that it's going to take fairly significant effort to make CA handle it gracefully. 

1. We can't just hand wave CA retrying deletion as a solution for LB blocking VM deletion. Maybe retrying deletion is the right answer for this part of the problem, but in that case we need someone to volunteer to do what may be pretty notable amount of work in CA (identifying all problematic mechanisms - I just named 3 off the top of my head but there may be more; fixing; testing all the edge cases).
2. One possible alternative is the delay-deletion CA annotation that was added in https://github.com/kubernetes/autoscaler/pull/2031. It seems to be a CA-specific annotation-based VM finalizer with timeout that was added for ingress controller (based on PR description anyway). I'm not aware of anyone actually using it and I'm not sure I like this solution, but it has the benefit of being already implemented.
3.  Finally, there is an interesting question: how is it working now? Do we just rely on node drain taking long enough for LB-controller to do any cleanup? It would seem logical, but I don't think it's the case. CA has a separate logic for deletng 'empty' nodes (empty ~= running only daemonsets) where CA doesn't actually wait for node drain. In this case I'd expect VM deletion to follow very quickly after tainting the node. The only explanations I see is that removing node from LBs is either super fast (I'm skeptical...) or cloud providers actually don't block VM deletions on LB. 

This is completely separate from the discussion on "how CA should signal LB" above, but I think equally important.

**Comment by alexanderConstantinescu** [comment]
> Do we just rely on node drain taking long enough for LB-controller to do any cleanup? It would seem logical, but I don't think it's the case. CA has a separate logic for deletng 'empty' nodes (empty ~= running only daemonsets) where CA doesn't actually wait for node drain. In this case I'd expect VM deletion to follow very quickly after tainting the node.

When you say "empty nodes" I don't suspect they're empty because all daemonsets are non-evictable - it's probably because they're system daemons? If so, system daemons never get any ingress traffic (through a Kubernetes load balancer service) do they? If they don't then we (Kubernetes) wouldn't have noticed it until now for that particular CA mode. But also: what we are discussing here is solving the problem of: "cluster ingress connectivity to a particular service doesn't benefit from connection draining". And instead, on node downscaling by the CA: all ingress connectivity to the application starts getting TCP timeouts instead of a graceful shutdown of their TCP connection. This is a problem space very formed around application/user pods, not system daemons like kube-proxy / etc which don't get ingress traffic. But no matter: not a lot of applications would notice/care about this failure mode. This discussion came out of a hypothetical scenario that me and @thockin had on another PR. That PR had a real issue. This one (though logically a real problem) I can imagine isn't easily observable by many humans....and probably only important for stateful applications (which most/all Kube pods aren't) for which a graceful shutdown might have a higher importance. 

I think that having the signal in t3 might just be a lucky coincidence, rather than something ingeniously planned. But I definitely think it has helped avoid the problem. We would have had more issues with non-GCed LB references to deleted VMs otherwise - unless the cloud providers clean those up too, if the service controller fails at doing that?

**Comment by thockin** [comment]
NOTE: I'm posting this as I wrote it, but the end sort of contradicts some of the other text.

TL;DR: I went and re-tested my assumptions, and I was WRONG.  At least on GCP, VM deletion automatically cleans up LBs.  I swear I tested it in the past and it failed, but I may have made a mistake.  CA is still over-simplistic in my opinion, and node lifecycle SHOULD be documented (IMO, in terms of "delete node" and finalizers), but maybe doesn't block here.


>  Our discussion here only assumes that the cloud provider will return an error if the VM is still referenced by some other construct_                            

If that is not true, there's not a whole lot we can do, right?  It could be broken and we can't really know...

> By placing the signal to the service controller at t5.5, I get the feeling we increase the risk of having the load balancer reference un-reaped - because the VM got deleted before the service controller even got a chance to try.

100% guaranteed

> we would need a signal in the other direction, i.e: the service controller which indicates that it has removed all references_

If you only consider LB, it seems reasonable.  But there's any number of similar things which MIGHT be needed.  The only way to do this correctly in general is to have interested parties indicate their interest and to delay the deletion until all of them have ACKed.

That sounds a lot like a finalizer to me.

> it doesn't ensure that all controllers act in any particular coordinated order

In general we don't offer that.  If you have 2 controllers which need to coordinate, that's on you.  As long as both keep retrying (and there's not a circular dep) it should resolve.

> If we use an annotation, CA will have to remember which annotated nodes it already deleted and avoid retrying VM deletion

That's an optimization, but not required, right?  If you retry a deletion, it should be idempotent - a deleted VM continues to be deleted :)

> Alternatively, instead of removing the finalizer, CA could try to delete the entire Node object itself, but then we'd need to make sure node controller isn't going to be unhappy about this.

It seems to me that "delete node" is the obviously correct approach, so there must be some reason we don't do it that way.  I asked @bobbypage  if there was a node lifecyle and the answer was approximately "no",

> I think it's much worse than that. In default configuration CA will completely block scale-down for 3 minutes after any failure.

OUCH

> This means that CA will not retry the same node for 10 minutes (and it will remove the taint on failure).

DOUBLE OUCH.  That seems pretty pessimistic - everything in k8s is async, and many failures are transient, and will auto-resolve if you just retry.

> To sum up: I don't think CA today is anywhere near being able to handle failures to remove nodes gracefully

This is disappointing to hear.  CA is an important component, but it's a problem if it blocks improvements in other subsystems...

> One possible alternative is the delay-deletion CA annotation that was added

This feels like a hack to me - we have a lifecycle mechanism in Kube APIs - why are we adding another layer on top?

> Finally, there is an interesting question: how is it working now?

As best I can tell - luck.  LB deletion can take a long time, but we do it "eagerly" today.  Too eagerly.  We're trading CA "simplicity" for doing a ton of extra work removing/adding nodes to the LB set and having a sort of pessimal LB experience.

We can make LB controller less eager in all cases EXCEPT for CA scale-down.  But if CA sets the taint and then aborts and removes it, we will add every node to every LB (which is what happens today).  It also means we will possible cause user-facing serving errors as LBs get yanked ungracefully.

> ... or cloud providers actually don't block VM deletions on LB

```
$ gcloud compute instances list
NAME                          ZONE           MACHINE_TYPE   PREEMPTIBLE  INTERNAL_IP  EXTERNAL_IP     STATUS
debian-buster                 us-central1-a  n1-standard-1               10.128.0.44                  TERMINATED
kubernetes-master             us-central1-b  n1-standard-1               10.128.0.59  34.27.8.27      RUNNING
kubernetes-minion-group-2qp7  us-central1-b  n1-standard-2               10.128.0.61  34.171.167.252  RUNNING
kubernetes-minion-group-73t5  us-central1-b  n1-standard-2               10.128.0.60  34.68.64.109    RUNNING
kubernetes-minion-group-74xn  us-central1-b  n1-standard-2               10.128.0.62  104.198.200.24  RUNNING
windows-1809                  us-central1-c  n1-standard-2               10.128.0.7                   TERMINATED
kind                          us-west1-b     e2-standard-4               10.138.0.4                   TERMINATED

$ gcloud compute forwarding-rules list
NAME                              REGION       IP_ADDRESS      IP_PROTOCOL  TARGET
a1787eb654d454ad7856c73fbeb274a3  us-central1  34.171.136.191  TCP          us-central1/targetPools/a1787eb654d454ad7856c73fbeb274a3

$ gcloud compute forwarding-rules describe a1787eb654d454ad7856c73fbeb274a3 --region us-central1
IPAddress: 34.171.136.191
IPProtocol: TCP
creationTimestamp: '2023-01-17T11:03:37.967-08:00'
description: '{"kubernetes.io/service-name":"default/iamyouare"}'
fingerprint: edLI2MAwaQs=
id: '2293937990112333798'
kind: compute#forwardingRule
labelFingerprint: 42WmSpB8rSM=
loadBalancingScheme: EXTERNAL
name: a1787eb654d454ad7856c73fbeb274a3
networkTier: PREMIUM
portRange: 80-80
region: https://www.googleapis.com/compute/v1/projects/thockin-dev/regions/us-central1
selfLink: https://www.googleapis.com/compute/v1/projects/thockin-dev/regions/us-central1/forwardingRules/a1787eb654d454ad7856c73fbeb274a3
target: https://www.googleapis.com/compute/v1/projects/thockin-dev/regions/us-central1/targetPools/a1787eb654d454ad7856c73fbeb274a3

$ gcloud compute target-pools describe a1787eb654d454ad7856c73fbeb274a3 --region us-central1
creationTimestamp: '2023-01-17T11:03:36.227-08:00'
description: '{"kubernetes.io/service-name":"default/iamyouare"}'
healthChecks:
- https://www.googleapis.com/compute/v1/projects/thockin-dev/global/httpHealthChecks/k8s-84ad1e081d11d529-node
id: '5408081114150162407'
instances:
- https://www.googleapis.com/compute/v1/projects/thockin-dev/zones/us-central1-b/instances/kubernetes-minion-group-2qp7
- https://www.googleapis.com/compute/v1/projects/thockin-dev/zones/us-central1-b/instances/kubernetes-minion-group-73t5
- https://www.googleapis.com/compute/v1/projects/thockin-dev/zones/us-central1-b/instances/kubernetes-master
- https://www.googleapis.com/compute/v1/projects/thockin-dev/zones/us-central1-b/instances/kubernetes-minion-group-74xn
kind: compute#targetPool
name: a1787eb654d454ad7856c73fbeb274a3
region: https://www.googleapis.com/compute/v1/projects/thockin-dev/regions/us-central1
selfLink: https://www.googleapis.com/compute/v1/projects/thockin-dev/regions/us-central1/targetPools/a1787eb654d454ad7856c73fbeb274a3
sessionAffinity: NONE

$ gcloud compute instances delete kubernetes-minion-group-2qp7 --zone us-central1-b
The following instances will be deleted. Any attached disks configured to be auto-deleted will be deleted unless they are attached to any other instances or the 
`--keep-disks` flag is given and specifies them for keeping. Deleting a disk is irreversible and any data on the disk will be lost.
 - [kubernetes-minion-group-2qp7] in [us-central1-b]

Do you want to continue (Y/n)?  y

Deleted [https://www.googleapis.com/compute/v1/projects/thockin-dev/zones/us-central1-b/instances/kubernetes-minion-group-2qp7].
```

Huh!  I *swear* this failed before.  At least in GCP, with the old-style (target-pools) API it does NOT fail.  I also tested it on the new-style API (instance-groups), and it was the same. 

Thanks for challenging that.  Maybe this is not a dire issue?  How could we get confirmation on other platforms?

I still think CA needs to be smarter about retrying.  Once it has made a decision, it seems like it should follow-through if the error is just a "reference count" sort of error.

Thoughts?

**Comment by alexanderConstantinescu** [comment]
Well, it seems [that PR mentioned](https://github.com/kubernetes/autoscaler/pull/2031): 

```
Functionality will be used by ingress controller to drain connections to the node before it is deleted. It also can be used by other components.
```

which fits quite nicely into what I just mentioned above. Maybe the service controller could use the same mechanism?

**Comment by thockin** [comment]
My point is that we HAVE a way to do that - finalizers.

**Comment by bobbypage** [comment]
This issue sounds like a good use case for a more standardized/documented node lifecycle, especially node termination "phase" that can be standardized across different controllers. 

Even if the LB deletion today works as tested above works, I'm sure there will be other usecases of resources that need to be cleaned up before the node deletion and having a standardized approach would be helpful I think. 

If we decided to formalize node lifecycle in k/k and went with the finalizer approach would there by any challenges for CA in adopting this? I believe the main contract that would need to be followed is that if a finalizer is set on the node object, CA should not delete the underlying instance from the cloud provider APIs until other some other controller would remove the finalizer.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/115139

**Comment by x13n** [comment]
I think formalized way of deleting nodes is the right way to go: without it we're risking adding hacks to the existing mechanism.

One major challenge I see here (compared with removing other resources) is that CA may not be able to delete the VM. For most resources, API objects represent the intent, which is then actuated by controllers. With nodes, the object doesn't represent an intent, it just reports the state of the underlying resource. Because of that, we need to think about some extra failure modes. What if CA doesn't have the permission to delete the VM? Or if there's some policy preventing this right now? Even if today existing forwarding rule doesn't block GCE VM deletion, this won't necessarily hold for other resources or other cloud providers.

**Comment by x13n** [comment]
> That's an optimization, but not required, right? If you retry a deletion, it should be idempotent - a deleted VM continues to be deleted :)

Yeah, that's an optimization, but I think an important one. In large clusters, the QPS from Cluster Autoscaler is massive already, even without this kind of retries. Once CA starts to hit API quotas, its performance will degrade, so I'd like us to design a mechanism that doesn't try to repeatedly delete an already deleted VM.

**Comment by alexanderConstantinescu** [comment]
For the record: I repeated @thockin's test above on AWS and Azure, and those clouds also allow the VM deletion to pass even though it's referenced by an LB. We thus wanted to get a discussion going with the different cloud-providers regarding whether or not we could remove this taint from the service controller's sync loop. So I filed https://github.com/kubernetes/kubernetes/pull/115204 as to get that discussion going there.

**Comment by MaciekPytel** [comment]
A lot of current implementation is due to historical reasons (pretty much all development in CA until very recently was focused on improving scale-up not scale-down) and there is certainly a lot of space for improvement.

That being said, I think there are tradeoffs here that should be considered:

1. I'd argue that the fact that scale-down can fail is a feature. CA's main job is to make sure that every pod can schedule and if we delete a node there is no guarantee we can add a new one later on (ex. because of cloud provider being out of stock). If a new pods gets scheduled leading to drain fail, than it's more correct for CA to abort scale-down than to risk making pods go pending. Similarly if CA gets restarted, we don't know what happened in the meantime and the original decision to scale-down may no longer be a correct one in the new cluster.
   * I don't think we can guarantee we'll always follow through on scale-down decision. The most likely place for scale-down to fail is in drain and it's also the most clear case for when we should abort (ex. if PDB doesn't allow disruption - this is simulated by CA, but ultimately validated server-side). Adding a new signal after drain (annotation, deletion timestamp) proposed above seems like a good way of reducing the number of cases where this happens.
   * Even a post-drain signal should not be seen as 100% guarantee that the node will eventually be removed - see below for cases on non-retriable cloud provider side errors. A large number of cloud provider side issues (quotas, ip space exhaustion, ASG-side max size limits, ...) leads to effective limit in node count and we may not be able to add new nodes before we delete old ones. If, for any reason, we can't delete a node and we don't give up we may be stuck in a situation where we have pending pods and tainted node for arbitrarily long time. 

2. As mentioned above the errors we're talking above are coming from cloud providers, not kubernetes. They're not standardized and often error modes are unique to specific provider. Some errors would go away on retry and some are persistent. Some examples:
     * A very common failure mode for CA is running out of cloud provider API call quota. Failing node deletion this way is technically a retriable error, but it can take minutes before a retry will succeed and there is a high risk we may not be able to add the node quickly (especially if you consider this usually happens when there are multiple kubernetes cluster and other services in the same project effectively competing for the quota). 
     * IIRC ASG will not let you remove a VM if it would bring it below min size of nodepool and setting different min size for CA and ASG is unfortunately a common user error (it's very easy to overlook..). That's a persistent error.
For scale-up, where CA has much more robust error handling we had to resort to cloud provider specific pattern matching on error messages (ie. regexps). As you can imagine the quality of error handling varies widely based on how actively cloudprovider integration of CA is handled and a lot of errors end up going through a generic 'unknown error' path.

Based on that experience, my opinion is that some retries on node deletion is a clear improvement (note that the number and frequency of retries need to be carefully tuned - those API call quotas really are a very common problem). At the same time we can't retry forever and if the error persists we'll have to assume it's persistent.

3. For autoscaler the latency/throughput is a key requirement. Today CA can add a 1000 nodes in a few minutes, but it can take many hours to scale-down the same number of nodes. We want to solve the LB-problem, but we shouldn't sacrifice (too much) scale-down speed/throughput to do so.

4. One problem I have with finalizers is that they're not time-bound and for all the reasons listed above CA needs to make a go/no go decision and remove the node or otherwise abandon scale-up attempt in O(minutes). Is there any precedent for time-bound finalizers?
Another one is that in almost any cluster CA is not the only component that removes nodes. Manual resizes and node upgrades are the first thing that come to mind. The challenge with those is that I don't think there is that everyone have their own implementation of them. All the ones I'm familiar with end up deleting VM and leaving node object cleanup to node controller. I don't see how we make everyone change their implementation, meaning those finalizers will only be respected if the node is deleted by CA.

5. I don't have very strong feelings here, but I see some benefits to NodeController and not CA deleting node object. As I wrote above CA is not the only thing deleting nodes. The logic of deleting node objects by NodeController is consistent across all those use-cases. The alternative is CA deleting nodes in some use-cases and NodeController in the other ones - less consistency, more codepaths.

**Comment by thockin** [comment]
@MaciekPytel

> CA's main job is to make sure that every pod can schedule and if we delete a node there is no guarantee we can add a new one later on

So the case you're thinking is:
* we should scale down
* drain
* delete kube Node (with finalizer)
* delete VM - fails
* time...
* we should scale up
* create VMs - fails (out of stock, quota, etc)

So now we have a Node that is attached to a VM which can't be deleted AND we can't get a new node.

Is that right?  It sounds pretty nefarious, but it also sounds like a pretty fundamental misconfiguration, and a situation that could arise for ANY managed resource, no?

> it's more correct for CA to abort scale-down than to risk making pods go pending.

This is a good argument for a well documented lifecycle - if we think this is really an important case, we should have a clear, unambiguous signal of intention to delete

> Similarly if CA gets restarted, we don't know what happened in the meantime and the original decision to scale-down may no longer be a correct one in the new cluster.

The problem is that there HAS TO come a time when the decision is made and is irreversible.  There are going to be controllers which need to act BEFORE a VM is deleted, and once they start, there's no coming back.  The simple answer is that this is "kubectl delete node" and VM deletion is a finalizer.  A more complicated answer is "some other annotation" which happens before VM delete which happens before node delete, but is still irrevocable in effect.

> I don't think we can guarantee we'll always follow through on scale-down decision. The most likely place for scale-down to fail is in drain and it's also the most clear case for when we should abort (ex. if PDB doesn't allow disruption - this is simulated by CA, but ultimately validated server-side). Adding a new signal after drain (annotation, deletion timestamp) proposed above seems like a good way of reducing the number of cases where this happens.

To be clear, I was suggesting that CA would:
* choose a node
* add a finalizer
* taint
* drain
* upon successful completion of drain, delete node
* finalizer deletes VM
* clear finalizer

In that model, drain can fail and the whole thing is aborted, but once drain succeeds, that node is deleted and there's no coming back.

> Even a post-drain signal should not be seen as 100% guarantee that the node will eventually be removed - see below for cases on non-retriable cloud provider side errors. A large number of cloud provider side issues (quotas, ip space exhaustion, ASG-side max size limits, ...) leads to effective limit in node count and we may not be able to add new nodes before we delete old ones. If, for any reason, we can't delete a node and we don't give up we may be stuck in a situation where we have pending pods and tainted node for arbitrarily long time.

We've been luck to not hit the above case, so far.  For example:

* hypothetical IPAM controller allocates IPs to a node and needs to release them before the VM can be deleted
* CA chooses victim node
* taint
* drain
* delete VM - fail

That will never resolve.  So we add a different annotation:

* hypothetical IPAM controller allocates IPs to a node and needs to release them before the VM can be deleted
* CA chooses victim node
* taint
* drain
* new annotation
* IPAm wakes up and starts releasing IPs
* delete VM - fails because IPAM is async
* remove new annotation

That VM is dead - those IPs are gone and not guaranteed to be recoverable.  I know CA *thinks* it is being clever, but you now have and undead machine which claims to be alive but isn't.

I am surprised we HAVEN'T hit this.

> For autoscaler the latency/throughput is a key requirement. Today CA can add a 1000 nodes in a few minutes, but it can take many hours to scale-down the same number of nodes. We want to solve the LB-problem, but we shouldn't sacrifice (too much) scale-down speed/throughput to do so.

I am not sure what the greater good is - error free serving or fast scale-down.

One problem I have with finalizers is that they're not time-bound and for all the reasons listed above CA needs to make a go/no go decision and remove the node or otherwise abandon scale-up attempt in O(minutes). Is there any precedent for time-bound finalizers?

What would you do if the time expires?  Finalizers are there to represent an essential cleanup that needs to happen.  Bypassing that could leak resources or far worse.

**Comment by thockin** [comment]
Anyway, it looks like we are unblocked on LB side, so ... punt to a later point?  I hate it but I know we're all busy.

**Comment by x13n** [comment]
I think the ability to drain nodes in parallel is a significant enough improvement (on average) to overall scale down performance to make regression from occassional VM deletion retries inconsequential. We can handle permanent errors like missing permissions by carefully configuring exponential backoffs (and use deletions batching, which was also implemented recently, to save even more precious qps).

> To be clear, I was suggesting that CA would:
> 
> * choose a node
> * add a finalizer
> * taint
> * drain
> * upon successful completion of drain, delete node
> * finalizer deletes VM
> * clear finalizer

This sounds like a very reasonable approach to me. The only thing I don't like about it is the fact that now both CA and Node controller are able to trigger Node object deletion. This might not be a problem in practice though, because deletion requests are idempotent, as you noted above.
